File: video-agent-test/internal/assets/assets.go
package assets

import (
	"fmt"
	"net/url"
	"os"
	"path/filepath"
	"strings"
)

type Ref struct {
	Value    string `json:"value"`
	Expected string `json:"expected"`
}

func IsHTTPURL(value string) bool {
	return strings.HasPrefix(value, "http://") || strings.HasPrefix(value, "https://")
}

func IsOSSURL(value string) bool {
	parsed, err := url.Parse(value)
	if err != nil {
		return false
	}
	host := strings.ToLower(parsed.Host)
	return strings.Contains(host, ".oss-") || strings.HasPrefix(host, "oss-") || strings.Contains(host, "aliyuncs.com")
}

func SidecarFor(path string) string {
	ext := filepath.Ext(path)
	if ext == "" {
		return path + ".url"
	}
	return strings.TrimSuffix(path, ext) + ext + ".url"
}

func ResolveLocalPath(ref string, projectRoot string) string {
	if filepath.IsAbs(ref) {
		return ref
	}
	return filepath.Join(projectRoot, ref)
}

func LocalRefToSidecarURL(ref string, projectRoot string) (urlValue string, sidecar string, err error) {
	localPath := ResolveLocalPath(ref, projectRoot)
	sidecar = SidecarFor(localPath)
	raw, err := os.ReadFile(sidecar)
	if err != nil {
		if os.IsNotExist(err) {
			return "", sidecar, nil
		}
		return "", sidecar, err
	}
	return strings.TrimSpace(string(raw)), sidecar, nil
}

func ResolveAssetURL(ref string, expected string, projectRoot string, allowNonOSS bool) (string, error) {
	ref = strings.TrimSpace(ref)
	if ref == "" {
		return "", fmt.Errorf("invalid %s reference: %q", expected, ref)
	}

	resolved := ref
	if !IsHTTPURL(ref) {
		urlValue, sidecar, err := LocalRefToSidecarURL(ref, projectRoot)
		if err != nil {
			return "", err
		}
		if urlValue == "" {
			return "", fmt.Errorf("%s asset is a local path without an OSS sidecar: %s\nUpload it first with scripts/bin/videoctl upload, then rewrite the frontmatter to the OSS URL, or create %s", expected, ref, sidecar)
		}
		resolved = urlValue
	}

	if !IsHTTPURL(resolved) {
		return "", fmt.Errorf("%s sidecar is not an http(s) URL: %s", expected, resolved)
	}
	if !allowNonOSS && !IsOSSURL(resolved) {
		return "", fmt.Errorf("%s URL is not recognized as an OSS URL: %s", expected, resolved)
	}
	return resolved, nil
}

func StringValue(value any) (string, bool) {
	if value == nil {
		return "", false
	}
	switch typed := value.(type) {
	case string:
		trimmed := strings.TrimSpace(typed)
		return trimmed, trimmed != ""
	default:
		raw := strings.TrimSpace(fmt.Sprint(typed))
		return raw, raw != ""
	}
}

func StringSlice(value any) []string {
	if value == nil {
		return nil
	}

	switch typed := value.(type) {
	case []string:
		return compactStrings(typed)
	case []any:
		items := make([]string, 0, len(typed))
		for _, item := range typed {
			if text, ok := StringValue(item); ok {
				items = append(items, text)
			}
		}
		return items
	case string:
		if strings.TrimSpace(typed) == "" {
			return nil
		}
		return []string{strings.TrimSpace(typed)}
	default:
		return nil
	}
}

func MapValue(value any) map[string]any {
	if value == nil {
		return nil
	}
	if typed, ok := value.(map[string]any); ok {
		return typed
	}
	return nil
}

func CollectImageRefs(frontmatter map[string]any) []Ref {
	var refs []Ref
	for _, field := range []string{"first_frame", "last_frame"} {
		if value, ok := StringValue(frontmatter[field]); ok {
			refs = append(refs, Ref{Value: value, Expected: "image"})
		}
	}

	assetMap := MapValue(frontmatter["assets"])
	for _, value := range StringSlice(assetMap["images"]) {
		refs = append(refs, Ref{Value: value, Expected: "image"})
	}

	if value, ok := StringValue(frontmatter["previous_frame_url"]); ok {
		refs = append(refs, Ref{Value: value, Expected: "image"})
	}

	return refs
}

func CollectVideoRefs(frontmatter map[string]any) []Ref {
	var refs []Ref
	assetMap := MapValue(frontmatter["assets"])
	for _, value := range StringSlice(assetMap["videos"]) {
		refs = append(refs, Ref{Value: value, Expected: "video"})
	}
	if value, ok := StringValue(frontmatter["previous_video_url"]); ok {
		refs = append(refs, Ref{Value: value, Expected: "video"})
	}
	return refs
}

func CollectValidationRefs(frontmatter map[string]any) []Ref {
	refs := append([]Ref{}, CollectImageRefs(frontmatter)...)
	refs = append(refs, CollectVideoRefs(frontmatter)...)
	return uniqueRefs(refs)
}

func compactStrings(values []string) []string {
	out := make([]string, 0, len(values))
	for _, value := range values {
		if trimmed := strings.TrimSpace(value); trimmed != "" {
			out = append(out, trimmed)
		}
	}
	return out
}

func uniqueRefs(refs []Ref) []Ref {
	seen := map[Ref]bool{}
	out := make([]Ref, 0, len(refs))
	for _, ref := range refs {
		if seen[ref] {
			continue
		}
		seen[ref] = true
		out = append(out, ref)
	}
	return out
}

File: agent/packages/opencode/src/video/assets.ts
import * as fs from "fs/promises"
import * as path from "path"

export interface AssetRef {
  value: string
  expected: "image" | "video"
}

export function isHTTPURL(value: string): boolean {
  return value.startsWith("http://") || value.startsWith("https://")
}

export function isOSSURL(value: string): boolean {
  try {
    const parsed = new URL(value)
    const host = parsed.host.toLowerCase()
    return host.includes(".oss-") || host.startsWith("oss-") || host.includes("aliyuncs.com")
  } catch {
    return false
  }
}

export function sidecarFor(filePath: string): string {
  return `${filePath}.url`
}

export function resolveLocalPath(ref: string, projectRoot: string): string {
  return path.isAbsolute(ref) ? ref : path.join(projectRoot, ref)
}

function uniqueLocalPathCandidates(ref: string, projectRoot: string, promptDir?: string): string[] {
  if (path.isAbsolute(ref)) return [ref]
  const candidates = [path.resolve(projectRoot, ref)]
  if (promptDir) candidates.push(path.resolve(promptDir, ref))
  return Array.from(new Set(candidates))
}

export async function localRefToSidecarURL(
  ref: string,
  projectRoot: string,
  promptDir?: string,
): Promise<{ url: string | undefined; sidecar: string; checked: string[] }> {
  const checked: string[] = []
  for (const localPath of uniqueLocalPathCandidates(ref, projectRoot, promptDir)) {
    const sidecar = sidecarFor(localPath)
    checked.push(sidecar)
    try {
      const raw = await fs.readFile(sidecar, "utf8")
      return { url: raw.trim(), sidecar, checked }
    } catch (err) {
      if ((err as NodeJS.ErrnoException).code === "ENOENT") continue
      throw err
    }
  }
  return { url: undefined, sidecar: checked[0] ?? sidecarFor(resolveLocalPath(ref, projectRoot)), checked }
}

export async function resolveAssetURL(
  ref: string,
  expected: string,
  projectRoot: string,
  allowNonOSS: boolean,
  promptDir?: string,
): Promise<string> {
  const trimmed = ref.trim()
  if (!trimmed) throw new Error(`invalid ${expected} reference: ${JSON.stringify(ref)}`)

  let resolved = trimmed
  if (!isHTTPURL(trimmed)) {
    const sidecar = await localRefToSidecarURL(trimmed, projectRoot, promptDir)
    if (!sidecar.url) {
      const hint = sidecar.checked.length > 1 ? sidecar.checked.join(" or ") : sidecar.sidecar
      throw new Error(
        `${expected} asset is a local path without an OSS sidecar: ${trimmed}. Upload it first or create ${hint}`,
      )
    }
    resolved = sidecar.url
  }

  if (!isHTTPURL(resolved)) throw new Error(`${expected} sidecar is not an http(s) URL: ${resolved}`)
  if (!allowNonOSS && !isOSSURL(resolved)) throw new Error(`${expected} URL is not recognized as an OSS URL: ${resolved}`)
  return resolved
}

export function stringValue(value: unknown): string | undefined {
  if (value === null || value === undefined) return undefined
  const text = typeof value === "string" ? value.trim() : String(value).trim()
  return text.length > 0 ? text : undefined
}

export function stringSlice(value: unknown): string[] {
  if (value === null || value === undefined) return []
  if (Array.isArray(value)) return value.map(stringValue).filter((item): item is string => item !== undefined)
  const single = stringValue(value)
  return single ? [single] : []
}

export function mapValue(value: unknown): Record<string, unknown> {
  if (value && typeof value === "object" && !Array.isArray(value)) return value as Record<string, unknown>
  return {}
}

export function collectImageRefs(frontmatter: Record<string, unknown>): AssetRef[] {
  const refs: AssetRef[] = []
  for (const field of ["first_frame", "last_frame"]) {
    const value = stringValue(frontmatter[field])
    if (value) refs.push({ value, expected: "image" })
  }
  const assetMap = mapValue(frontmatter.assets)
  for (const value of stringSlice(assetMap.images)) refs.push({ value, expected: "image" })
  const previous = stringValue(frontmatter.previous_frame_url)
  if (previous) refs.push({ value: previous, expected: "image" })
  return refs
}

export function collectVideoRefs(frontmatter: Record<string, unknown>): AssetRef[] {
  const refs: AssetRef[] = []
  const assetMap = mapValue(frontmatter.assets)
  for (const value of stringSlice(assetMap.videos)) refs.push({ value, expected: "video" })
  const previous = stringValue(frontmatter.previous_video_url)
  if (previous) refs.push({ value: previous, expected: "video" })
  return refs
}

export function collectValidationRefs(frontmatter: Record<string, unknown>): AssetRef[] {
  const seen = new Set<string>()
  const out: AssetRef[] = []
  for (const ref of [...collectImageRefs(frontmatter), ...collectVideoRefs(frontmatter)]) {
    const key = `${ref.expected}\0${ref.value}`
    if (seen.has(key)) continue
    seen.add(key)
    out.push(ref)
  }
  return out
}
